### PR TITLE
Implemented An ActionVisitor Class

### PR DESCRIPTION
**Pull request by garethellis0**
### Description
- Added an `ActionVisitor` class and all associated required methods in `Action` subclasses.
- Cleaned up some comments in other visitor classes on the way

**NOTE:** This is branched off of #1020, the diff should be _significantly_ smaller once that has merged.

### Testing Done
Unit tests pass.

### Resolved Issues
Part of #773

### Length Justification
This _should_ end up under 500 lines once 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/ai/hl/stp/action/action_visitor.h [hunk 1]
```diff
@@ -0,0 +1,40 @@
+#pragma once
+
+// We forward-declare all the actions because if we include them we induce a
+// circular dependency between the Individual library for each action and this
```
Comment (akhilveeraghanta) [inline src/software/ai/hl/stp/action/action_visitor.h]: nit
```suggestion
// circular dependency between the individual library for each action and this
```

File: src/software/ai/hl/stp/action/all_actions.h [hunk 2]
```diff
@@ -0,0 +1,10 @@
+#pragma once
+
+#include "software/ai/hl/stp/actions/chip_action.h"
```
Comment (akhilveeraghanta) [inline src/software/ai/hl/stp/action/all_actions.h:3]: can this file just be the output of a genrule so we have less things to worry about when adding a new action?

File: src/software/ai/hl/stp/action/all_actions.h [hunk 3]
```diff
@@ -0,0 +1,10 @@
...
+#include "software/ai/hl/stp/actions/dribble_action.h"
+#include "software/ai/hl/stp/actions/intercept_ball_action.h"
+#include "software/ai/hl/stp/actions/kick_action.h"
+#include "software/ai/hl/stp/actions/move_action.h"
+#include "software/ai/hl/stp/actions/movespin_action.h"
+#include "software/ai/hl/stp/actions/picot_action.h"
```
Comment (akhilveeraghanta) [inline src/software/ai/hl/stp/action/all_actions.h]:
```suggestion
#include "software/ai/hl/stp/actions/pivot_action.h"
```

**Review by akhilveeraghanta — CHANGES_REQUESTED** [review]
Overall LGTM, the `action_visitor.h` file looks really clean, just a few comments

**Comment by jonathanlew** [comment]
Can you merge before I review?

File: src/software/ai/hl/stp/action/chip_action.h [hunk 4]
```diff
@@ -18,49 +19,44 @@ class ChipAction : public Action
...
-     * where the robot will meet the ball as it's moving. We need to be able to specify
-     * where the chip will take place even if the ball isn't there yet, which is why the
-     * chip_origin is separate. The ball is used for other calculations, such as when
-     * the ball has been chipped and the action is done.
+     * The chip origin is given (instead of just using the ball position), so that we
+     * can line up chips where the ball isn't present yet. For example, specifying where
+     * a chip will take place where the robot will meet the ball as it's moving. We need
+     * to be able to specify where the chip will take place even if the ball isn't there
+     * yet, which is why the chip_origin is separate. The ball is used for other
+     * calculations, such as when the ball has been chipped and the action is done.
```
Comment (matthewberends) [inline src/software/ai/hl/stp/action/chip_action.h]: Is this comment still accurate. The ball isn't being passed in anymore so how is it being used for calculations?

**Review by matthewberends — COMMENTED** [review]
This is looking good. Just a minor documentation comment.

**Comment by garethellis0** [comment]
@jonathanlew - merged, should be good for review now!

File: src/software/ai/hl/stp/action/chip_action.h (hunk 4 above)
Comment (garethellis0) [reply src/software/ai/hl/stp/action/chip_action.h]: Ah, the ball is being passed in via `updateWorldParams` above. Hope that clears things up, but please let me know if not, documenting this clearly is super important.

**Review by jonathanlew — DISMISSED** [review]
Looks good, nothing to add beyond what others have brought up!

File: src/software/ai/hl/stp/action/all_actions.h (hunk 2 above)
Comment (MathewMacDougall) [reply src/software/ai/hl/stp/action/all_actions.h:3]: I think @jonathanlew brought that up separately. @jonathanlew is there a ticket for this? (I didn't see one)

I believe we would address this separately (would need the same for tactics, primitives, and really anything with a visitor), but that can be @garethellis0 's call

**Review by MathewMacDougall — COMMENTED** [review]
Nothing to add beyond the existing comments. Looks good.

File: src/software/ai/hl/stp/action/all_actions.h (hunk 2 above)
Comment (garethellis0) [reply src/software/ai/hl/stp/action/all_actions.h:3]: Yep, #865 is open for this, seems out-of-scope for this PR.

File: src/software/ai/hl/stp/action/action_visitor.h (hunk 1 above)
Comment (garethellis0) [reply src/software/ai/hl/stp/action/action_visitor.h]: changed as suggested!

File: src/software/ai/hl/stp/action/all_actions.h (hunk 3 above)
Comment (garethellis0) [reply src/software/ai/hl/stp/action/all_actions.h]: changed!

File: src/software/ai/hl/stp/action/chip_action.h (hunk 4 above)
Comment (matthewberends) [reply src/software/ai/hl/stp/action/chip_action.h]: Oh ok misses that. Yeah that clears it up. Thanks

**Review by MathewMacDougall — APPROVED** [review]
LGTM

**Comment by garethellis0** [comment]
@matthewberends @jonathanlew - could you please re-approve?

**Review by jonathanlew — APPROVED** [review]

**Review by matthewberends — APPROVED** [review]

**Review by akhilveeraghanta — APPROVED** [review]
LGTM :shipit:

**Pull request merged**